src/cargo/core/workspace.rs

```diff
@@ -1841,3 +1825,106 @@ impl InheritableFields {
             })
     }
 }
+
+fn parse_manifest(manifest_path: &Path, config: &Config) -> CargoResult<EitherManifest> {
+    let key = manifest_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (manifest, _nested_paths) = read_manifest(manifest_path, source_id, config)?;
+    Ok(manifest)
+}
+
+pub fn find_workspace_root(manifest_path: &Path, config: &Config) -> CargoResult<Option<PathBuf>> {
+    find_root_iter(manifest_path, config)
+        .find_map(|ances_manifest_path| {
+            let manifest = parse_manifest(&ances_manifest_path, config);
+            match manifest {
+                Ok(manifest) => match *manifest.workspace_config() {
+                    WorkspaceConfig::Root(ref ances_root_config) => {
+                        debug!("find_root - found a root checking exclusion");
+                        if !ances_root_config.is_excluded(manifest_path) {
+                            debug!("find_root - found!");
+                            Some(Ok(ances_manifest_path))
+                        } else {
+                            None
+                        }
+                    }
+                    WorkspaceConfig::Member {
+                        root: Some(ref path_to_root),
+                    } => {
+                        debug!("find_root - found pointer");
+                        Some(Ok(read_root_pointer(&ances_manifest_path, path_to_root)))
+                    }
+                    WorkspaceConfig::Member { .. } => None,
+                },
+                Err(e) => Some(Err(e)),
+            }
+        })
+        .transpose()
+}
+
+fn read_root_pointer(member_manifest: &Path, root_link: &str) -> PathBuf {
+    let path = member_manifest
+        .parent()
+        .unwrap()
+        .join(root_link)
+        .join("Cargo.toml");
+    debug!("find_root - pointer {}", path.display());
+    paths::normalize_path(&path)
+}
+
+fn find_root_iter<'a>(
+    manifest_path: &'a Path,
+    config: &'a Config,
+) -> impl Iterator<Item = PathBuf> + 'a {
+    LookBehind::new(paths::ancestors(manifest_path, None).skip(2))
+        .into_iter()
```

Comment:
Is `into_iter` needed?  Seems like we should already have an iterator

Reply:
It's not needed, I'll remove it.